### PR TITLE
Restores Spell Card targetting behaviour

### LINKED ISSUE #52946: spell cards still don't home in on people
The spell's supposed to make a shotgun spray of cards that'll turn mid-flight to home in on people, but all it does currently is make a shotgun spray of shitty cards that don't home in on people. As such, the spell does basically jack shit to people unless you're only a tile or so away from them (and even then, the damage output of the spell cards is so poor that you need multiple "reloads"(/recastings) of the spell to kill someone at point blank range with it).

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

Fixes #52946

This has been broken ever since #44112 which apparently removed the proc in this component which selected a target for performance reasons and just... didn't replace it with anything else? So it's been bricked ever since.

In order to restore the removed mouse tracking behaviour I implemented the pattern used in scoped weapons and the kinesis module. As this was the third thing I could find to use this code, I abstracted it out into an object where most of the shared behaviour lives. I tested those things too and they still seem to do what they used to.

Here it is in action:

https://user-images.githubusercontent.com/7483112/221954852-22244bb1-7c87-452d-a9b0-ebed81c4c1ef.mp4

Because this spell now applies a full screen overlay in order to function, it's really begging for someone to make some touhou-style art with cards and patterns around the edge of the screen while you are in aiming mode, but I'm not going to be the one to make it.

I tidied this component up and refactored it as best I could while I was reworking it to use a full screen overlay but I am... not totally confident that it should be a component at all given that it keeps being created and destroyed. But also it has worked that way for like four years now, so who am I to say.

Oh yeah also the icon for this spell broke, so I fixed it.

## Why It's Good For The Game

Makes a reasonably terrible wizard spell marginally less bad.
Maybe now that this component works, some other things can use it?
If we make a fourth thing which follows your cursor it won't need to copy/paste code around.

## Changelog

:cl:
fix: Spell Cards from the Wizard spell will now home in somewhat on the target nearest to your cursor.
fix: The Spell Cards spell now displays the correct icon instead of a big red "error" text.
/:cl:

### CODE REVIEW

File: code/datums/components/lock_on_cursor.dm [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+#define LOCKON_RANGING_BREAK_CHECK if(current_ranging_id != this_id){return LOCKON_IGNORE_RESULT}
+
+/**
+ * ### Lock on Cursor component
+ *
+ * Finds the nearest targets to your cursor and passes them into a callback, also drawing an icon on top of them.
+ */
+/datum/component/lock_on_cursor
+	dupe_mode = COMPONENT_DUPE_ALLOWED
+	// Appearance to overlay onto whatever we are targetting
```
Comment (SuperNovaa41) [inline code/datums/components/lock_on_cursor.dm:11]:
```suggestion
	/// Appearance to overlay onto whatever we are targetting
```

File: code/datums/components/lock_on_cursor.dm [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	var/atom/movable/screen/fullscreen/cursor_catcher/lock_on/mouse_tracker
+	/// Ranging ID for some kind of tick check safety calculation
+	var/current_ranging_id = 0
+
+/datum/component/lock_on_cursor/Initialize(range = 5, list/typecache = list(), amount = 1, list/immune = list(), icon = 'icons/mob/silicon/cameramob.dmi', icon_state = "marker", datum/callback/when_locked, datum/callback/target_callback)
+	if(!ismob(parent))
+		return COMPONENT_INCOMPATIBLE
+	if (amount < 1 || range < 0)
+		CRASH("Invalid range or amount argument")
+	lock_cursor_range = range
```
Comment (SuperNovaa41) [inline code/datums/components/lock_on_cursor.dm:39]: https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#when-passing-vars-through-new-or-initializes-arguments-use-srcvar

Use the same variable names for stuff like this.
i.e.
```
src.lock_cursor_range = lock_cursor_range
```
etc for all variables that apply in this proc

File: code/datums/components/lock_on_cursor.dm [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+			if(target_turf)
+				targets |= special_list_filter(target_turf.contents, can_target_callback)
+				if(targets.len >= lock_amount)
+					targets.Cut(lock_amount+1)
+					return targets
+		LOCKON_RANGING_BREAK_CHECK
+		cd++
+		CHECK_TICK
+
+/atom/movable/screen/fullscreen/cursor_catcher/lock_on
```
Comment (SuperNovaa41) [inline code/datums/components/lock_on_cursor.dm:194]: you don't need to remove this but you don't actually need this, the atom will be defined with just the proc

File: code/game/objects/effects/cursor_catcher.dm [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	/// Pixel y we send to the scope component.
+	var/given_y
+	/// The turf we send to the scope component.
+	var/turf/given_turf
+	/// Mouse parameters, for calculation.
+	var/mouse_params
+
+/// Links this up with a mob
+/atom/movable/screen/fullscreen/cursor_catcher/proc/assign_to_mob(mob/new_owner)
+	owner = new_owner
```
Comment (SuperNovaa41) [inline code/game/objects/effects/cursor_catcher.dm:21]:
```suggestion
/atom/movable/screen/fullscreen/cursor_catcher/proc/assign_to_mob(mob/owner)
	src.owner = owner
```

**Review by SuperNovaa41 — CHANGES_REQUESTED** [review]

File: code/datums/components/lock_on_cursor.dm (hunk 3 above)
Comment (Jacquerel) [reply code/datums/components/lock_on_cursor.dm:194]: I just think it's slightly neater to define explicitly anyway.

File: code/modules/spells/spell_types/pointed/spell_cards.dm [hunk 5]
```diff
@@ -42,11 +42,11 @@
...
-		/datum/component/lockon_aiming, \
-		range = 5, \
-		typecache = GLOB.typecache_living, \
-		amount = 1, \
-		when_locked = CALLBACK(src, PROC_REF(on_lockon_component)))
+		/datum/component/lock_on_cursor, \
+		lock_cursor_range = 5, \
+		target_typecache = GLOB.typecache_living, \
+		lock_amount = 1, \
+		on_lock = CALLBACK(src, PROC_REF(on_lockon_component)))
```
Comment (san7890) [inline code/modules/spells/spell_types/pointed/spell_cards.dm:49]:
```suggestion
		on_lock = CALLBACK(src, PROC_REF(on_lockon_component)), \
	)
```

let's make it trailing

File: code/datums/components/lock_on_cursor.dm (hunk 3 above)
Comment (san7890) [reply code/datums/components/lock_on_cursor.dm:194]: if you're going to do it like that can you dmdoc the difference between this and parent

File: code/datums/components/lock_on_cursor.dm [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	/// Callback to call when we have decided on our targets, is passed the list of final targets
+	var/datum/callback/on_lock
+	/// Callback to call in order to validate a potential target
+	var/datum/callback/can_target_callback
+	/// Full screen overlay which is used to track mouse position
+	var/atom/movable/screen/fullscreen/cursor_catcher/lock_on/mouse_tracker
+	/// Ranging ID for some kind of tick check safety calculation
+	var/current_ranging_id = 0
+
+/datum/component/lock_on_cursor/Initialize(lock_cursor_range = 5, list/target_typecache = list(), lock_amount = 1, list/immune = list(), icon = 'icons/mob/silicon/cameramob.dmi', icon_state = "marker", datum/callback/on_lock, datum/callback/can_target_callback)
```
Comment (san7890) [inline code/datums/components/lock_on_cursor.dm:34]: might be worth to multiline this?

File: code/datums/components/scope.dm [hunk 7]
```diff
@@ -29,20 +29,20 @@
 	))
 
 /datum/component/scope/process(delta_time)
-	if(!tracker.marksman.client)
-		stop_zooming(tracker.marksman)
+	if(!tracker.owner.client)
```
Comment (san7890) [inline code/datums/components/scope.dm:32]: can we cache this to some proc-local variable so we don't have to keep accessing two different things in `process()`? if you're scared of caching the client due to disconnect/BYOND memes, that's fine, but at least cache `tracker.owner`. will make it easier to read for the rest of the proc imo

**Review by san7890 — CHANGES_REQUESTED** [review]
small nits

**Pull request merged**